### ISSUE TITLE
High Level m3u8 interface

### ISSUE THREAD

**Issue by Luro02** [sile/hls_m3u8#29]
It'd be nice to have some kind of high level struct, that does all the parsing, for example to download chunks and decrypt them correctly. For downloading one should use [reqwest](https://github.com/seanmonstar/reqwest) and for decrypting [aes](https://github.com/RustCrypto/block-ciphers/).

The API could look something like this;
``` rust
use hls_m3u8::Client;

let client = Client::new(&input_str); // internally reqwest is used
client.with_request(|req| { // this function will be applied to each request.
    // this can be used to add additional headers to the request
    // (some servers could for example require cookies)
    req
});

client.with_custom(|url| { // -> impl Future<Output = Result<Vec<u8>, E>>
   // returns the raw data from the url
    async {
        // do async reqwest here...
    }
});

{
    let cache = client.get_mut_cache() -> &mut Cache;
    cache.set_size(1024*1024*1024); // maximum cached packages are 1GiB
    cache.set_max_packages(50);
    cache.clear(); // empties the cache
} // drop &mut Cache

// Chunks should be loaded in the background, depending on the configuration
// The next chunk can be fetched like this
client.next().await?; // returns next chunk, either from the cache or directly downloaded
```

**Comment by Luro02** [comment]
I think this should be put in a separate crate that is built on top of this one. Maybe calles `hls_m3u8-client`?

**Comment by gquintard** [comment]
Hi,

Sorry if it's the wrong issue to comment on, but would be nice to be able to get back an `enum` telling what kind of manifest this is. Or maybe it already exist and I didn't find it?

**Comment by Luro02** [comment]
>  would be nice to be able to get back an enum telling what kind of manifest this is.

I agree, before I knew anything about m3u8 files I had a similar thought, but now I am pretty certain that you will most likely get a `MasterPlaylist` first (the links to the `MediaPlaylist` files are inside the `MasterPlaylist`)

>  Or maybe it already exist and I didn't find it?

No such a method does not currently exist.

> Sorry if it's the wrong issue to comment on,

I made an issue for it, see #58
